### PR TITLE
[SPARK-41165][CONNECT] Avoid hangs in the arrow collect code path

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
Two changes:
1. Make sure connect's arrow result path properly deals with errors, and avoids hangs.
2. Fix a common source of non-serializable exceptions in `SparkConnectStreamHandler`.

### Why are the changes needed?
The current Arrow result code path for connect assumes no error can happen during execution. As a result it will hang when an error occurs.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added a UT.

### CODE REVIEW

**Comment by zhengruifeng** [comment]
also cc @cloud-fan @amaliujia

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectServiceSuite.scala [hunk 1]
```diff
@@ -55,4 +65,38 @@ class SparkConnectServiceSuite extends SharedSparkSession {
           && schema.getFields(1).getType.getKindCase == proto.DataType.KindCase.STRING)
     }
   }
+
+  test("failures in the arrow collect path should not cause hangs") {
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectServiceSuite.scala:69]: nit
```suggestion
  test("SPARK-41165: failures in the arrow collect path should not cause hangs") {
```

logic-wise, this is what I would do too. Looks good to me.

**Review by cloud-fan — APPROVED** [review]

**Comment by amaliujia** [comment]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**